pkg/safe/routine.go

```diff
@@ -59,12 +59,23 @@ func (p *Pool) GoCtx(goroutine routineCtx) {
 	p.routinesCtx = append(p.routinesCtx, goroutine)
 	p.waitGroup.Add(1)
 	Go(func() {
+		defer p.waitGroup.Done()
 		goroutine(p.ctx)
-		p.waitGroup.Done()
 	})
 	p.lock.Unlock()
 }
 
+// AddGo adds a recoverable goroutine, and can be stopped with stop chan
+func (p *Pool) AddGo(goroutine func(stop chan bool)) {
```

Comment:
If AddGo is not used in any other place than in the tests, let's make it non-exported please.